src/azstoragetorch/io.py

```diff
@@ -0,0 +1,188 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import io
+import os
+from typing import get_args, Optional, Union, Literal
+import urllib.parse
+
+from azure.identity import DefaultAzureCredential
+from azure.core.credentials import (
+    AzureSasCredential,
+    TokenCredential,
+)
+
+
+_SUPPORTED_MODES = Literal["rb"]
+_SDK_CREDENTIAL_TYPE = Optional[
+    Union[
+        AzureSasCredential,
+        TokenCredential,
+    ]
+]
+_AZSTORAGETORCH_CREDENTIAL_TYPE = Union[_SDK_CREDENTIAL_TYPE, Literal[False]]
+
+
+class BlobIO(io.IOBase):
```

Comment:
> Is the concept of different blob types relevant here? 

Possibly. If we did want to support different blob types for writes (for the initial implementation, it was going to be block blob support only), I'd first suggest exploring accepting the blob_type as part of the `BlobIO` constructor e.g.:
```python
BlobIO('<blob-url>', 'wb', blob_type='page')
```
The `BlobIO` would have the same file-like object interface methods. But under the hood, we could delegate to different uploader implementations based on what type of blob was specified and we could keep the write logic needed for each blob type to their own uploader class. It would be sort of similar approach to how the Python SDK has different class implementations through a [common chunk uploader interface](https://github.com/Azure/azure-sdk-for-python/blob/7b18366f341b2b85ee10397387315a3e3bec40c4/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py#L120).

There is probably some other interfaces we can explore here such as having explicit `BlobIO` classes based on the blob type (e.g. `PageBlobIO`) or even exploring if it makes sense to map the `mode` to a blob type (e.g. `ab` would make it an append blob). Nonetheless, I think there are ways we should be able to keep the file-like methods the same as the standard library but still be able to specify a blob type if we decide we need it.

Thoughts?

Reply:
Gotcha, this all makes sense and should not stop any further developments, as this initial interface doesn't interfere with introducing blob types later on.

I agree that a lightweight initial implementation with just block blob support is a good start so we can get deep on getting it working fully E2E, then we can revisit and fold in concepts like blob types and adjust accordingly.